flatsurf/geometry/hyperbolic.py

```diff
@@ -0,0 +1,774 @@
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation, either version 2 of the License, or
+#  (at your option) any later version.
+#
+#  sage-flatsurf is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with sage-flatsurf. If not, see <https://www.gnu.org/licenses/>.
+######################################################################
+
+from sage.structure.parent import Parent
+from sage.structure.element import Element
+from sage.structure.unique_representation import UniqueRepresentation
+
+
+class HyperbolicPlane(Parent, UniqueRepresentation):
+    r"""
+    The hyperbolic plane over a base ring.
+
+    All objects in the plane must be specified over the given base ring. Note
+    that, in some representations, objects might appear to live in a larger
+    ring. E.g., when specifying a line by giving a center and the square of
+    its radius in the half plane model, then the ideal endpoints of this line
+    might have coordinates in the ring after adjoining a square root.
+
+    The implemented elements of the plane are convex subsets such as (finite
+    and infinite) points, geodesics, closed half planes, and closed convex
+    polygons.
+
+    ALGORITHM:
+
+    We do not use a fixed representation of the hyperbolic plane internally but
+    switch between the Poincaré half plane and the Klein model freely.
+
+    For the Klein model, we use a unit disc centered at (0, 0). The map from
+    the Poincaré half plane sends the imaginary unit `i` to the center at the
+    origin, and sends 0 to (0, -1), 1 to (1, 0), -1 to (-1, 0) and infinity to
+    (0, 1). The Möbius transformation
+
+    .. MATH::
+
+        z \mapsto \frac{z-i}{1 - iz}
+
+    maps from the half plane model to the Poincaré disc model. We then post-compose this with the map that goes from the Poincaré disc model to the Klein disc model, which in polar coordinates sends
+
+    .. MATH::
+
+        (\phi, r)\mapsto \left(\phi, \frac{2r}{1 + r^2}\right).
+
+    When we write this map out explicitly in Euclidean coordinates, we get
+
+    .. MATH::
+
+        (x, y) \mapsto \frac{1}{1 + x^2 + y^2}\left(2x, -1 + x^2 + y^2\right)
+
+    and
+
+    .. MATH::
+
+        (x, y) \mapsto \frac{1}{1 - y}\left(x, \sqrt{1 - x^2 - y^2}\right),
+
+    for its inverse.
+
+    A geodesic in the Poincaré half plane is then given by an equation of the form
+
+    .. MATH::
+
+        a(x^2 + y^2) + bx + c = 0
+
+    which converts to an equation in the Klein disc as
+
+    .. MATH::
+
+        (a + c) - bx + (a - c)y = 0.
+
+    Note that the intersection of two geodesics with coefficients in a field
+    `K` therefore has coordinates in the same field `K` in either model.
```

Comment:
I am not sure I understand correctly this remark. More precisely the "either model" part: in the Poincaré half plane, the intersection of two geodesics is generally not in the ground field.

Reply:
You're right. That statement survived from the previous wrong transformation.